polkadot/xcm/xcm-builder/src/asset_exchange.rs

```diff
@@ -0,0 +1,146 @@
+use sp_runtime::traits::Zero;
+use sp_std::vec;
+use xcm::prelude::*;
+use xcm_executor::{
+	traits::{AssetExchange, MatchesFungibles},
+	AssetsInHolding,
+};
+
+/// An adapter from [`pallet_asset_conversion::SwapCredit`] and
+/// [`pallet_asset_conversion::QuoteExchangePrice`] to [`xcm_executor::traits::AssetExchange`].
+///
+/// Takes just one fungible asset in `give` and allows only one fungible asset in `want`.
+pub struct FungiblesPoolAdapter<AssetConversion, Fungibles, Matcher, AccountId>(
+	PhantomData<(AssetConversion, Fungibles, Matcher, AccountId)>,
+);
+impl<AssetConversion, Fungibles, Matcher, AccountId> AssetExchange
+	for FungiblesPoolAdapter<AssetConversion, Fungibles, Matcher, AccountId>
+where
+	AssetConversion: SwapCredit<
+			AccountId,
+			Balance = u128,
+			AssetKind = Fungibles::AssetId,
+			Credit = fungibles::Credit<AccountId, Fungibles>,
+		> + QuoteExchangePrice<Balance = u128, AssetKind = Fungibles::AssetId>,
+	Fungibles: fungibles::Balanced<AccountId, Balance = u128>,
+	Matcher: MatchesFungibles<Fungibles::AssetId, Fungibles::Balance>,
+{
+	fn exchange_asset(
+		_: Option<&Location>,
+		give: AssetsInHolding,
+		want: &Assets,
+		_: bool,
+	) -> Result<AssetsInHolding, AssetsInHolding> {
+		let give_asset = give.fungible_assets_iter().next().ok_or_else(|| {
+			log::trace!(
+				target: "xcm::FungiblesPoolAdapter::exchange_asset",
+				"No fungible asset was in `give`.",
+			);
+			give.clone()
+		})?;
+		let want_asset = want.get(0).ok_or_else(|| {
+			log::trace!(
+				target: "xcm::FungiblesPoolAdapter::exchange_asset",
+				"No asset was in `want`.",
+			);
+			give.clone()
+		})?;
+		let (give_asset_id, balance) =
+			Matcher::matches_fungibles(&give_asset).map_err(|error| {
+				log::trace!(
+					target: "xcm::FungiblesPoolAdapter::exchange_asset",
+					"Could not map XCM asset {:?} to FRAME asset. Error: {:?}",
+					give,
+					error,
+				);
+				give.clone()
+			})?;
+		let (want_asset_id, want_amount) =
+			Matcher::matches_fungibles(&want_asset).map_err(|error| {
+				log::trace!(
+					target: "xcm::FungiblesPoolAdapter",
+					"Could not map XCM asset {:?} to FRAME asset. Error: {:?}",
+					want,
+					error,
+				);
+				give.clone()
+			})?;
+
+		// We have to do this to convert the XCM assets into credit the pool can use.
+		let swap_asset = give_asset_id.clone().into();
+		let credit_in = Fungibles::issue(give_asset_id, balance);
+		log::trace!(target: "xcm", "Credit in: {:?}", credit_in.peek());
+
+		// Do the swap.
+		let (credit_out, credit_change) =
+			<AssetConversion as SwapCredit<_>>::swap_tokens_for_exact_tokens(
+				vec![swap_asset, want_asset_id],
+				credit_in,
+				want_amount,
+			)
+			.map_err(|(credit_in, _)| {
+				drop(credit_in);
+				give.clone()
+			})?;
+
+		log::trace!(target: "xcm", "Credit out: {:?}, Credit change: {:?}", credit_out.peek(), credit_change.peek());
+		debug_assert!(credit_change.peek() == Zero::zero());
```

Comment:
not sure if we should require a client to know the exact `give` amount. this impl looks very specific to the case we wanna handle with this PR.